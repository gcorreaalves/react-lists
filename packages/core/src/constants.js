export const PACKAGE_NAME = 'react-chat'
