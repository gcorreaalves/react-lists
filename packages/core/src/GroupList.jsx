import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles'
import Core from './Core'
import ListItem from './ListItem'
import ListIndex from './ListIndex'
import { PACKAGE_NAME } from './constants'

const Group = props => {
  const { items, itemRenderer, listHeight, showGroupIndex, onItemClick } = props
  const listRef = React.useRef(null)

  const handleIndexItemClick = id => e => {
    e.preventDefault()

    if (listRef.current) {
      listRef.current.scrollTop = document.querySelector(id).offsetTop
    }
  }

  const renderItem = (item, index) => {
    const key = item.id || index

    return (
      <ListItem
        key={key}
        item={item}
        renderer={itemRenderer}
        onClick={onItemClick}
      />
    )
  }

  const renderItems = () => {
    const groupKey = 'header'

    return Object.keys(items).map(key => {
      const header = items[key][groupKey]

      if (header) {
        return (
          <li id={`${PACKAGE_NAME}-index-${key.toLowerCase()}`} key={key}>
            <p style={styles.groupedListHeader}>{header}</p>
            <ul style={styles.groupedList}>
              {items[key].data.map(renderItem)}
            </ul>
          </li>
        )
      } else {
        throw new Error(
          `The object should have a attribute called "${groupKey}"`
        )
      }
    })
  }

  return (
    <div style={styles.root}>
      <Core listRef={listRef} {...props}>
        {renderItems()}
      </Core>
      {showGroupIndex && (
        <ListIndex
          height={listHeight}
          items={Object.keys(items)}
          onItemClick={handleIndexItemClick}
        />
      )}
    </div>
  )
}

Group.propTypes = {
  displayLoading: PropTypes.bool,
  items: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.shape({
      data: PropTypes.arrayOf(PropTypes.object),
      header: PropTypes.string
    })
  ]).isRequired,
  itemRenderer: PropTypes.func.isRequired,
  listHeight: PropTypes.number.isRequired,
  loadingRenderer: PropTypes.func,
  reversed: PropTypes.bool,
  showGroupIndex: PropTypes.bool,
  onBottomReached: PropTypes.func,
  onItemClick: PropTypes.func,
  onLoadMore: PropTypes.func,
  onTopReached: PropTypes.func
}

export default Group
