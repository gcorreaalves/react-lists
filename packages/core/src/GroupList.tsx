import React from 'react'
import Core, { CoreTypes } from './Core'
import ListItem from './ListItem'
import ListIndex from './ListIndex'
import { PACKAGE_NAME } from './constants'

const styles: { [key: string]: React.CSSProperties } = {
  root: {
    position: 'relative'
  },
  groupedList: {
    listStyle: 'none',
    padding: 0,
    margin: 0,
    width: '100%',
    display: 'flex',
    flexDirection: 'column-reverse'
  },
  subHeader: {
    marginBottom: '10px',
    paddingBottom: '5px',
    top: 0,
    width: '100%',
    zIndex: 1
  },
  subHeaderSticky: {
    position: 'sticky'
  }
}

interface GroupListTypes extends CoreTypes {
  displayIndexMenu?: boolean
  displaySubHeaders?: boolean
  items: {
    [key: string]: {
      header: string
      data: Array<unknown>
    }
  }
  itemRenderer: React.FunctionComponent
  stickySubHeaders: boolean
  subHeaderRenderer?: React.FunctionComponent<{ text: string }>
  onItemClick?: (event: React.MouseEvent) => void
}

const SubHeaderDefault: React.FunctionComponent<{ text: string }> = ({
  text
}) => <>{text}</>

const GroupList: React.FunctionComponent<GroupListTypes> = props => {
  const {
    displayIndexMenu,
    items,
    itemRenderer,
    listHeight,
    subHeaderRenderer,
    onItemClick,
    displaySubHeaders = true,
    stickySubHeaders = true
  } = props
  const groupKey = 'header'
  const listRef = React.useRef(null)

  const handleIndexItemClick = id => e => {
    e.preventDefault()

    if (listRef.current) {
      listRef.current.scrollTop = document.querySelector(id).offsetTop
    }
  }

  const renderItem = (item, index) => {
    const key = item.id || index

    return (
      <ListItem
        key={key}
        item={item}
        renderer={itemRenderer}
        onClick={onItemClick}
      />
    )
  }

  const renderItems = () => {
    return Object.keys(items).map(key => {
      const header = items[key][groupKey]
      const SubHeaderRenderer = subHeaderRenderer || SubHeaderDefault
      let subHeaderStyles = styles.subHeader

      if (stickySubHeaders) {
        subHeaderStyles = { ...subHeaderStyles, ...styles.subHeaderSticky }
      }

      if (header) {
        return (
          <li id={`${PACKAGE_NAME}-index-${key.toLowerCase()}`} key={key}>
            {displaySubHeaders && (
              <header
                className={`${PACKAGE_NAME}-subheader`}
                style={subHeaderStyles}
              >
                <SubHeaderRenderer text={header} />
              </header>
            )}
            <ul style={styles.groupedList}>
              {items[key].data.map(renderItem)}
            </ul>
          </li>
        )
      } else {
        throw new Error(
          `The object should have a attribute called "${groupKey}"`
        )
      }
    })
  }

  return (
    <div style={styles.root}>
      <Core listRef={listRef} {...props}>
        {renderItems()}
      </Core>
      {displayIndexMenu && (
        <ListIndex
          height={listHeight}
          items={Object.keys(items).map(key => items[key][groupKey])}
          onItemClick={handleIndexItemClick}
        />
      )}
    </div>
  )
}

export default GroupList
