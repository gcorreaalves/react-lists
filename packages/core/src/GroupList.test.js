import React from 'react'
import { mount, render } from 'enzyme'
import GroupList from './GroupList'
import { PACKAGE_NAME } from './constants'

describe('GroupList component', () => {
  let mockMenu = {}

  beforeEach(() => {
    mockMenu = {
      clothing: {
        header: 'Clothing',
        data: ['Man', 'Woman', 'Boy', 'Girl', 'Baby', 'Shoes', 'Hats']
      },
      eletronics: {
        header: 'Eletronics',
        data: ['Computer', 'Smartphone', 'Tablet', 'TV', 'Games']
      },
      sports: {
        header: 'Sports',
        data: ['Fishing', 'Football', 'Golf', 'Tennis']
      }
    }

    const div = document.createElement('div')
    div.setAttribute('id', 'container')
    document.body.appendChild(div)
  })

  it('should render correctly', () => {
    const wrapper = mount(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render 3 grouped items', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
      />
    )

    expect(wrapper.find(`.${PACKAGE_NAME}-subheader`).length).toBe(3)
  })

  it('should display sub headers', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
      />
    )

    expect(wrapper.find(`#${PACKAGE_NAME}-index-clothing`).length).toBe(1)
    expect(wrapper.find(`#${PACKAGE_NAME}-index-eletronics`).length).toBe(1)
    expect(wrapper.find(`#${PACKAGE_NAME}-index-sports`).length).toBe(1)
  })

  it('should not display sub headers', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
        displaySubHeaders={false}
      />
    )

    expect(wrapper.find(`.${PACKAGE_NAME}-subheader`).length).toBe(0)
  })

  it('should render custom sub headers', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
        subHeaderRenderer={({ text }) => (
          <div className='custom-subheader'>{text}</div>
        )}
      />
    )

    expect(wrapper.find('.custom-subheader').length).toBe(3)
    expect(wrapper.find('.custom-subheader').first().text()).toBe('Clothing')
  })

  it('should throw an Error if there is no header', () => {
    delete mockMenu.clothing.header

    expect(() => {
      render(
        <GroupList
          listHeight={400}
          items={mockMenu}
          itemRenderer={({ item }) => item}
        />
      )
    }).toThrowError(
      new Error('The object should have a attribute called "header"')
    )
  })

  it('should display index menu', () => {
    const wrapper = render(
      <GroupList
        displayIndexMenu
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
      />
    )

    expect(wrapper.find('nav').length).toBe(1)
    expect(wrapper.find('nav').find('li').length).toBe(3)
    expect(wrapper.find('nav').find('li:first').text()).toBe('Clothing')
  })

  it('should not display index menu', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
      />
    )

    expect(wrapper.find('nav').length).toBe(0)
  })

  it('should not make sub headers sticky', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
        stickySubHeaders={false}
      />
    )

    expect(
      wrapper.find(`.${PACKAGE_NAME}-subheader`).first().prop('style')
    ).not.toHaveProperty('position', 'sticky')
  })

  it('should make sub headers sticky', () => {
    const wrapper = render(
      <GroupList
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
        stickySubHeaders
      />
    )

    expect(
      wrapper.find(`.${PACKAGE_NAME}-subheader`).first().prop('style')
    ).toHaveProperty('position', 'sticky')
  })

  it('should scroll list when clicking on index menu', () => {
    const spy = jest.spyOn(document, 'querySelector')

    const wrapper = mount(
      <GroupList
        displayIndexMenu
        listHeight={400}
        items={mockMenu}
        itemRenderer={({ item }) => item}
      />,
      { attachTo: document.getElementById('container') }
    )

    wrapper.find('nav').find('li').first().find('a').simulate('click')

    expect(spy).toBeCalledWith(`#${PACKAGE_NAME}-index-clothing`)
  })

  afterEach(() => {
    const div = document.getElementById('container')

    if (div) {
      document.body.removeChild(div)
    }

    jest.clearAllMocks()
  })
})
