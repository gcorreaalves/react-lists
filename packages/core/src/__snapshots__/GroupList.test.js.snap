// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupList component should render correctly 1`] = `
<GroupList
  displayIndexMenu={false}
  displayLoading={false}
  displaySubHeaders={true}
  itemRenderer={[Function]}
  items={
    Object {
      "clothing": Object {
        "data": Array [
          "Man",
          "Woman",
          "Boy",
          "Girl",
          "Baby",
          "Shoes",
          "Hats",
        ],
        "header": "Clothing",
      },
      "eletronics": Object {
        "data": Array [
          "Computer",
          "Smartphone",
          "Tablet",
          "TV",
          "Games",
        ],
        "header": "Eletronics",
      },
      "sports": Object {
        "data": Array [
          "Fishing",
          "Football",
          "Golf",
          "Tennis",
        ],
        "header": "Sports",
      },
    }
  }
  listHeight={400}
  stickySubHeaders={true}
>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <Core
      displayIndexMenu={false}
      displayLoading={false}
      displaySubHeaders={true}
      itemRenderer={[Function]}
      items={
        Object {
          "clothing": Object {
            "data": Array [
              "Man",
              "Woman",
              "Boy",
              "Girl",
              "Baby",
              "Shoes",
              "Hats",
            ],
            "header": "Clothing",
          },
          "eletronics": Object {
            "data": Array [
              "Computer",
              "Smartphone",
              "Tablet",
              "TV",
              "Games",
            ],
            "header": "Eletronics",
          },
          "sports": Object {
            "data": Array [
              "Fishing",
              "Football",
              "Golf",
              "Tennis",
            ],
            "header": "Sports",
          },
        }
      }
      listHeight={400}
      listRef={
        Object {
          "current": <ul
            style="list-style: none; padding: 0px; position: relative; margin: 0px; width: 100%; height: 400px; overflow-y: scroll;"
          >
            <li
              id="react-lists-index-clothing"
            >
              <header
                class="react-lists-subheader"
                style="margin-bottom: 10px; padding-bottom: 5px; top: 0px; width: 100%; z-index: 1; position: sticky;"
              >
                Clothing
              </header>
              <ul
                style="list-style: none; padding: 0px; margin: 0px; width: 100%; display: flex; flex-direction: column-reverse;"
              >
                <li>
                  Man
                </li>
                <li>
                  Woman
                </li>
                <li>
                  Boy
                </li>
                <li>
                  Girl
                </li>
                <li>
                  Baby
                </li>
                <li>
                  Shoes
                </li>
                <li>
                  Hats
                </li>
              </ul>
            </li>
            <li
              id="react-lists-index-eletronics"
            >
              <header
                class="react-lists-subheader"
                style="margin-bottom: 10px; padding-bottom: 5px; top: 0px; width: 100%; z-index: 1; position: sticky;"
              >
                Eletronics
              </header>
              <ul
                style="list-style: none; padding: 0px; margin: 0px; width: 100%; display: flex; flex-direction: column-reverse;"
              >
                <li>
                  Computer
                </li>
                <li>
                  Smartphone
                </li>
                <li>
                  Tablet
                </li>
                <li>
                  TV
                </li>
                <li>
                  Games
                </li>
              </ul>
            </li>
            <li
              id="react-lists-index-sports"
            >
              <header
                class="react-lists-subheader"
                style="margin-bottom: 10px; padding-bottom: 5px; top: 0px; width: 100%; z-index: 1; position: sticky;"
              >
                Sports
              </header>
              <ul
                style="list-style: none; padding: 0px; margin: 0px; width: 100%; display: flex; flex-direction: column-reverse;"
              >
                <li>
                  Fishing
                </li>
                <li>
                  Football
                </li>
                <li>
                  Golf
                </li>
                <li>
                  Tennis
                </li>
              </ul>
            </li>
          </ul>,
        }
      }
      stickySubHeaders={true}
    >
      <div
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <ul
          onScroll={[Function]}
          style={
            Object {
              "height": 400,
              "listStyle": "none",
              "margin": 0,
              "overflowY": "scroll",
              "padding": 0,
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <li
            id="react-lists-index-clothing"
            key="clothing"
          >
            <header
              className="react-lists-subheader"
              style={
                Object {
                  "marginBottom": "10px",
                  "paddingBottom": "5px",
                  "position": "sticky",
                  "top": 0,
                  "width": "100%",
                  "zIndex": 1,
                }
              }
            >
              <Component
                text="Clothing"
              >
                Clothing
              </Component>
            </header>
            <ul
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column-reverse",
                  "listStyle": "none",
                  "margin": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
            >
              <ListItem
                item="Man"
                key="0"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Man"
                  >
                    Man
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Woman"
                key="1"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Woman"
                  >
                    Woman
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Boy"
                key="2"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Boy"
                  >
                    Boy
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Girl"
                key="3"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Girl"
                  >
                    Girl
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Baby"
                key="4"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Baby"
                  >
                    Baby
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Shoes"
                key="5"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Shoes"
                  >
                    Shoes
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Hats"
                key="6"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Hats"
                  >
                    Hats
                  </itemRenderer>
                </li>
              </ListItem>
            </ul>
          </li>
          <li
            id="react-lists-index-eletronics"
            key="eletronics"
          >
            <header
              className="react-lists-subheader"
              style={
                Object {
                  "marginBottom": "10px",
                  "paddingBottom": "5px",
                  "position": "sticky",
                  "top": 0,
                  "width": "100%",
                  "zIndex": 1,
                }
              }
            >
              <Component
                text="Eletronics"
              >
                Eletronics
              </Component>
            </header>
            <ul
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column-reverse",
                  "listStyle": "none",
                  "margin": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
            >
              <ListItem
                item="Computer"
                key="0"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Computer"
                  >
                    Computer
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Smartphone"
                key="1"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Smartphone"
                  >
                    Smartphone
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Tablet"
                key="2"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Tablet"
                  >
                    Tablet
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="TV"
                key="3"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="TV"
                  >
                    TV
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Games"
                key="4"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Games"
                  >
                    Games
                  </itemRenderer>
                </li>
              </ListItem>
            </ul>
          </li>
          <li
            id="react-lists-index-sports"
            key="sports"
          >
            <header
              className="react-lists-subheader"
              style={
                Object {
                  "marginBottom": "10px",
                  "paddingBottom": "5px",
                  "position": "sticky",
                  "top": 0,
                  "width": "100%",
                  "zIndex": 1,
                }
              }
            >
              <Component
                text="Sports"
              >
                Sports
              </Component>
            </header>
            <ul
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column-reverse",
                  "listStyle": "none",
                  "margin": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
            >
              <ListItem
                item="Fishing"
                key="0"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Fishing"
                  >
                    Fishing
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Football"
                key="1"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Football"
                  >
                    Football
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Golf"
                key="2"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Golf"
                  >
                    Golf
                  </itemRenderer>
                </li>
              </ListItem>
              <ListItem
                item="Tennis"
                key="3"
                renderer={[Function]}
              >
                <li
                  onClick={[Function]}
                >
                  <itemRenderer
                    item="Tennis"
                  >
                    Tennis
                  </itemRenderer>
                </li>
              </ListItem>
            </ul>
          </li>
        </ul>
      </div>
    </Core>
  </div>
</GroupList>
`;
